File: app/src/main/java/fr/nihilus/xenobladechronicles/monsters/add/AddMonsterViewModel.java
package fr.nihilus.xenobladechronicles.monsters.add;

import android.arch.core.util.Function;
import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.MutableLiveData;
import android.arch.lifecycle.Transformations;
import android.arch.lifecycle.ViewModel;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import javax.inject.Inject;

import fr.nihilus.xenobladechronicles.model.Area;
import fr.nihilus.xenobladechronicles.model.Monster;
import fr.nihilus.xenobladechronicles.monsters.MonsterRepository;
import fr.nihilus.xenobladechronicles.util.AbsentLiveData;

/**
 * Un {@link ViewModel} gérant la tâche d'ajout d'un {@link Monster}.
 */
public class AddMonsterViewModel extends ViewModel {

    private final MonsterRepository mRepository;
    private MutableLiveData<String> mSearchQuery = new MutableLiveData<>();
    private LiveData<String[]> mKindSearchResults;

    @Inject
    AddMonsterViewModel(final MonsterRepository repository) {
        mRepository = repository;
        mKindSearchResults = Transformations.switchMap(mSearchQuery, new Function<String, LiveData<String[]>>() {
            @Override
            public LiveData<String[]> apply(String search) {
                if (search == null || search.trim().isEmpty()) {
                    return AbsentLiveData.create();
                } else {
                    return mRepository.getKinds(search);
                }
            }
        });
    }

    /**
     * Spécifie les caractères saisis par l'utilisateur lorsqu'il précise l'espèce
     * à laquelle le monstre ajouté appartient.
     * Les suggestions retournées par {@link #getKinds()} seront automatiquement mises à jour.
     */
    void setKindQuery(@NonNull String input) {
        if (input.equals(mSearchQuery.getValue())) {
            return;
        }

        mSearchQuery.setValue(input);
    }

    /**
     * @return suggestions pour la saisie de l'espèce à laquelle un monstre appartient
     */
    LiveData<String[]> getKinds() {
        return mKindSearchResults;
    }

    void addMonster(int level, @NonNull String name, @NonNull Area area, @Nullable String location,
                    @Nullable String kind) {
        Monster newMonster = new Monster();
        newMonster.setName(name);
        newMonster.setLevel(level);
        newMonster.setArea(area);
        newMonster.setLocation(location);
        newMonster.setKind(kind);
        newMonster.setDefeated(false);

        mRepository.add(newMonster);
    }


}

File: app/src/main/java/fr/nihilus/xenobladechronicles/monsters/add/AddMonsterViewModel.kt
package fr.nihilus.xenobladechronicles.monsters.add

import androidx.lifecycle.LiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import androidx.lifecycle.viewModelScope
import fr.nihilus.xenobladechronicles.datastore.MonsterDao
import fr.nihilus.xenobladechronicles.model.Area
import fr.nihilus.xenobladechronicles.model.Monster
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.time.ExperimentalTime
import kotlin.time.milliseconds

/**
 * Un [ViewModel] gérant la tâche d'ajout d'un [Monster].
 */
class AddMonsterViewModel @Inject constructor(
    private val monsterDao: MonsterDao
) : ViewModel() {

    private val searchQuery = MutableStateFlow("")

    /**
     * @return suggestions pour la saisie de l'espèce à laquelle un monstre appartient
     */
    @OptIn(ExperimentalTime::class, FlowPreview::class)
    val kinds: LiveData<List<String>> = searchQuery
        .debounce(300.milliseconds)
        .flatMapLatest { rawSearch ->
            val search = rawSearch.trim()
            when {
                search.isEmpty() -> flowOf(emptyList())
                else -> monsterDao.searchKinds(rawSearch)
            }
        }
        .asLiveData()

    /**
     * Spécifie les caractères saisis par l'utilisateur lorsqu'il précise l'espèce
     * à laquelle le monstre ajouté appartient.
     * Les suggestions retournées par [.getKinds] seront automatiquement mises à jour.
     */
    fun setKindQuery(input: String) {
        searchQuery.value = input
    }

    fun registerMonster(level: Int, name: String, kind: String, area: Area, location: String) {
        viewModelScope.launch {
            val monster = Monster(id = null, level, name, kind, area, location, isDefeated = false)
            monsterDao.insert(monster)
        }
    }
}
